### ISSUE TITLE
Revise Documentation Structure

### ISSUE THREAD

**Issue by kotfu** [python-cmd2/cmd2#709]
Related to #572, I propose an overhaul of the documentation. I intend for this issue to be the place where we discuss and agree on a high level structure and some guiding principles for the documentation. The writing of the various parts which don't yet exist can be handled in other issues.

## Guiding Principles

- We have gone to great lengths to retain compatibility with the standard library `cmd`, the documentation should make it easy for developers to understand how to move from `cmd` to `cmd2`, and what benefits that will provide
- We should provide both descriptive and reference documentation.
- API reference documentation should be generated from docstrings in the code
- Documentation should include rich hyperlinking to other areas of the documentation, and to the API reference

## High Level Structure

Here's a proposed top level outline for the documentation. I'm not proposing and order yet, just an unsorted list of high level topics.

- Overview
    - what cmd2 is, major features for users and developers, and what it isn't
    - Quick Start Tutorial - walk through of construction of a simple cmd2 based application
- Installation and System Requirements
- Migrating from cmd in the standard library
    - What you have to change
    - What you might have to change
    - What you get for free
    - What features you'll probably want to take advantage of first
- Features
    - Generating output - output, errors, paging, 
    - Argument parsing
    - History
    - The Prompt
    - Startup Commands
    - Shortcuts, Aliases, and Macros
    - Settings - current settings and adding your own
    - Completion
    - Creating Help
    - Working with the operating system - redirecting output, executing commands, editors, paging, etc.
    - Multi-line commands
    - Color
    - Enabling and disabling commands
    - Using the clipboard
    - Python shell and scripting
    - Transcripts
    - Plugins and Hooks
- HOWTO's - A list of recipes/examples for accomplishing common or difficult things, similar to the examples we provide in the source code distribution
    - getting command line arguments from the operating system into your cmd2 program
    - creating a custom help system
    - creating a custom settings system
    - multi-threaded operations
- API Reference (every public function and attribute)

**Comment by xNinjaKittyx** [comment]
I think having a quickstart (giving one of the simple examples in this repo), would be helpful for people completely new to the library (and cmd in general).

**Comment by kotfu** [comment]
Love it, added it to the outline above.

**Comment by tleonhardt** [comment]
Overall this looks quite good. 

A few sections I would recommend adding include:
* How cmd2 compares and contrasts to other CLI alternatives like click and prompt-toolkit
* Scripting (separate sections because very different)
    * Text file scripting with `run_script`
    * Python scripting with `run_pyscript`
* Interactive Python shells (separate from scripting because use case is different)
    * Python shell with `py`
    * Opt-in **IPython** shell with `ipy`
* Automation and calling cmd2 from other CLI/CLU tools via commands at invocation and quit
* Feedback to user from asynchronous commands/events

**Comment by kotfu** [comment]
Agree with @tleonhardt suggestions too. Some of these things we already have in our current documentation, like items 1, 4, and 5. We just need to put them in the right place(s). For example, the async feedback thing could go in the "Prompt" section, as well as having it's own spot in the HOWTO section.

I've created a branch with much of this structure implemented, I'll get it filled out with a placeholder for each topic, and then if we like the overall structure, I can start writing the various sections.

**Comment by kotfu** [comment]
I've pushed a few more commits on this branch. I think I have a placeholder for everything that should be documented, with the exception of all of the examples. I also cleaned up the Getting Started section, and while that section is not content complete, I think it represents how I envision each section coming together.

I also started a documentation conventions document.

I think we need to keep this branch until we have at least:

- the overview section completed
- the migrating section completed
- at least one example documented
- the existing documentation all parceled out into the various feature stub pages

Then we can merge it, which will result in master having at least as good of documentation as it does now. We can write about the rest of the features and examples in future issues and branches.
